kernel/objects/capability-spinning/objects/RevokeOperation.cc

```diff
@@ -99,98 +99,92 @@ namespace mythos {
+        root->finishRevoke();
+        RETURN(Error::SUCCESS);
+      }
+      ASSERT(leafCap.isZombie());
+      if (leafCap.getPtr() == _guarded) {
+        leaf->unlock_cap();
+        // attempted to delete guarded object
+        THROW(Error::CYCLIC_DEPENDENCY);
+      }
+      leaf->lock_cap();
+      if (leaf->cap() != leafCap) {
+        MLOG_DETAIL(mlog::cap, "leaf cap changed concurrently");
+        leaf->unlock_cap();
+        continue;
+      }
+      auto delRes = leafCap.getPtr()->deleteCap(*leaf, leafCap, *this);
+      auto prevres = leaf->lock_prev();
+      ASSERT_MSG(prevres, "somebody unlinked CapEntry currently in unlinking process");
+      leaf->lock_next();
+      if (delRes) {
+        leaf->unlinkAndUnlockLinks();
+        leaf->reset();
+      } else {
+        // deletion failed
+        // Either tried to delete a portal that is currently deleting
+        // or tried to to delete _guarded via a recursive call.
+        leaf->unlock_cap();
+        RETHROW(delRes);
+      }
     } while (leaf != root);
     // deleted root
     RETURN(Error::SUCCESS);
   }
 
-  bool RevokeOperation::_startTraversal(CapEntry* root, Cap rootCap)
-  {
-    if (!root->lock_prev()) {
-      // start is currently unlinked
-      // must be the work of another deleter ... success!
-      return false;
-    }
-    if (root->prev() == root) {
-      // this is the actual root of the tree
-      // and has no children
-      // avoid deadlocks and finish the operation
-      root->unlock_prev();
-      return false;
-    }
-    root->lock();
-    // compare only value, not the zombie state
-    if (root->cap().asZombie() != rootCap.asZombie()) {
-      // start has a new value
-      // must be the work of another deleter ... success!
-      root->unlock();
-      root->unlock_prev();
-      return false;
-    }
-    return true;
-  }
-
-  CapEntry* RevokeOperation::_findLockedLeaf(CapEntry* root)
+  // not sure if we even need that locking
+  bool RevokeOperation::_findLeaf(CapEntry* const root, Cap const rootCap, CapEntry*& leafEntry, Cap& leafCap)
   {
-    auto curEntry = root;
+    MLOG_ERROR(mlog::cap, __PRETTY_FUNCTION__, DVAR(root) );
+    leafEntry = root;
+    leafCap = rootCap;
     while (true) {
-      auto curCap = curEntry->cap();
-      auto nextEntry = curEntry->next();
-      // wait for potencially allocated cap to become usable/zombie
-      Cap nextCap;
-      for (nextCap = nextEntry->cap();
-          nextCap.isAllocated();
-          nextCap = nextEntry->cap()) {
-        ASSERT(!nextEntry->isDeleted());
-        hwthread_pause();
+      auto nextEntry = leafEntry->next();
+      if (nextEntry) {
+        Cap nextCap;
+        // wait for potencially allocated cap to become usable/zombie
+        for (nextCap = nextEntry->cap();
+            nextCap.isAllocated();
+            nextCap = nextEntry->cap()) {
+          ASSERT(!nextEntry->isDeleted());
+          hwthread_pause();
```

Comment:
shouldn't we recheck our own next pointer in the case that reading the successors capability failed?

Reply:
Why? As long as nobody deletes the parent (not possible at the moment if this is a child), I don't see how that could go wrong. It does not matter if somebody added a child before us, and since the entry is killed (most likely) ... good luck doing that. We can only become not-a-leaf if sb. adds a child to us.

Reply:
Our next pointer can change because of the successor being deleted concurrently. This is allowed because nothing was locked during our traversal to the first leaf.